File: frontend/src/types/requests.ts
/**
 * Types for the contract intake / Requests surface.
 *
 * Mirrors `backend/app/schemas/requests.py`. Status, request_type,
 * contract_type, and priority are deliberately string-typed so the
 * UI can display whatever values the backend actually returns;
 * recommended values are listed inline.
 */
export type ContractRequestStatus =
  | "open"
  | "in_progress"
  | "completed"
  | "cancelled";

export interface ContractRequest {
  id: string;
  organization_id: string;
  title: string;
  description: string | null;
  /** Suggested: new_contract, review_existing, amendment, renewal, other. */
  request_type: string | null;
  /** Suggested: NDA, MSA, SOW, DPA, Employment Agreement, Lease, Other. */
  contract_type: string | null;
  status: ContractRequestStatus | string;
  /** Suggested: low, normal, high, urgent. */
  priority: string | null;
  requester_name: string | null;
  requester_email: string | null;
  counterparty_name: string | null;
  /** ISO date (YYYY-MM-DD). */
  due_date: string | null;
  assigned_to: string | null;
  linked_contract_id: string | null;
  linked_template_id: string | null;
  created_at: string;
  updated_at: string;
  created_by: string | null;
  metadata_json: Record<string, unknown> | null;
}

export interface ContractRequestCreateRequest {
  title: string;
  description?: string | null;
  request_type?: string | null;
  contract_type?: string | null;
  priority?: string | null;
  requester_name?: string | null;
  requester_email?: string | null;
  counterparty_name?: string | null;
  due_date?: string | null;
  assigned_to?: string | null;
  linked_contract_id?: string | null;
  linked_template_id?: string | null;
  metadata_json?: Record<string, unknown> | null;
}

export interface ContractRequestUpdateRequest {
  title?: string;
  description?: string | null;
  request_type?: string | null;
  contract_type?: string | null;
  status?: ContractRequestStatus | string;
  priority?: string | null;
  requester_name?: string | null;
  requester_email?: string | null;
  counterparty_name?: string | null;
  due_date?: string | null;
  assigned_to?: string | null;
  linked_contract_id?: string | null;
  linked_template_id?: string | null;
  metadata_json?: Record<string, unknown> | null;
}

export interface ListContractRequestFilters {
  status?: string;
  request_type?: string;
  contract_type?: string;
  priority?: string;
  assigned_to?: string;
  /** ISO date (YYYY-MM-DD). */
  due_before?: string;
  /** ISO date (YYYY-MM-DD). */
  due_after?: string;
  include_cancelled?: boolean;
}

/**
 * Body for ``POST /api/requests/{id}/convert-to-contract``.
 *
 * Mirrors ``AgreementGenerationRequest`` — the conversion endpoint
 * reuses the template generation service. ``variable_values`` is
 * keyed by the linked template's variable ``key``s.
 */
export interface ConvertRequestToContractRequest {
  title?: string | null;
  variable_values: Record<string, unknown>;
}

/**
 * Response shape for ``POST /api/requests/{id}/convert-to-contract``.
 *
 * Reuses the same contract / artifact / markdown_snapshot projections
 * the templates surface returns, plus the freshly-updated request row
 * so the UI can swap both pieces of state in one shot. Storage and
 * encryption fields never appear here.
 */
export interface ConvertRequestToContractResponse {
  request: ContractRequest;
  contract: {
    id: string;
    title: string;
    status: string;
    mime_type: string;
    file_hash_sha256: string;
    page_count: number | null;
    created_at: string;
    updated_at: string;
  };
  artifact: {
    id: string;
    contract_id: string;
    artifact_type: "generated_docx" | string;
    storage_backend: string;
    filename: string | null;
    mime_type: string | null;
    file_hash_sha256: string | null;
    size_bytes: number | null;
    source: "template_generation" | string | null;
    is_official: boolean;
    created_at: string;
    metadata_json: Record<string, unknown> | null;
  };
  markdown_snapshot: {
    id: string;
    contract_id: string;
    markdown_text: string;
    source_kind: string;
    converter_name: string;
    converter_version: string | null;
    conversion_status: string;
    conversion_warnings: unknown[] | null;
    created_at: string;
  } | null;
  variables_used: string[];
}

/**
 * Body for ``POST /api/requests/{id}/convert-upload`` (PR #65).
 *
 * The endpoint is multipart/form-data: the API client builds a
 * ``FormData`` from this shape so the request includes the file part
 * plus the optional metadata fields below as form values.
 */
export interface ConvertRequestUploadInput {
  file: File;
  /** Optional Contract title; falls back to the derived filename. */
  title?: string | null;
  counterparty_name?: string | null;
  contract_type?: string | null;
  notes?: string | null;
  signal?: AbortSignal;
}

/**
 * Response shape for ``POST /api/requests/{id}/convert-upload``.
 *
 * Same projections used by the template-conversion response, minus
 * ``variables_used`` (there are no template variables on this path).
 * Storage / encryption fields are excluded by construction at the
 * schema layer.
 */
export interface ConvertRequestUploadResponse {
  request: ContractRequest;
  contract: {
    id: string;
    title: string;
    status: string;
    mime_type: string;
    file_hash_sha256: string;
    page_count: number | null;
    created_at: string;
    updated_at: string;
  };
  artifact: {
    id: string;
    contract_id: string;
    artifact_type: "original_upload" | string;
    storage_backend: string;
    filename: string | null;
    mime_type: string | null;
    file_hash_sha256: string | null;
    size_bytes: number | null;
    source: "request_upload" | string | null;
    is_official: boolean;
    created_at: string;
    metadata_json: Record<string, unknown> | null;
  };
  markdown_snapshot: {
    id: string;
    contract_id: string;
    markdown_text: string;
    source_kind: string;
    converter_name: string;
    converter_version: string | null;
    conversion_status: string;
    conversion_warnings: unknown[] | null;
    created_at: string;
  } | null;
}

File: backend/app/schemas/requests.py
"""Request/response schemas for ``/api/requests``.

``ContractRequest`` is the intake/business workflow object — the place a
non-lawyer says "I need an NDA with X" so the legal team can triage,
assign, and convert later. The schema deliberately keeps the taxonomy
(``request_type``, ``contract_type``, ``priority``) free-form so users
can model their own categories without a migration; suggested values are
documented inline.
"""
from __future__ import annotations

import uuid
from datetime import date, datetime
from typing import Any

from pydantic import BaseModel, ConfigDict, Field

from app.schemas.artifacts import ContractArtifactResponse
from app.schemas.contract_intake import (
    DuplicateContractCandidateResponse,
    ExtractedContractMetadataResponse,
)
from app.schemas.contracts import ContractListItemResponse
from app.schemas.markdown import ContractMarkdownSnapshotResponse


class ContractRequestCreateRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    title: str = Field(min_length=1, max_length=255)
    description: str | None = None

    # Suggested: new_contract, review_existing, amendment, renewal, other.
    request_type: str | None = Field(default=None, max_length=64)
    # Suggested: NDA, MSA, SOW, DPA, Employment Agreement, Lease, Other.
    contract_type: str | None = Field(default=None, max_length=64)
    # Suggested: low, normal, high, urgent.
    priority: str | None = Field(default=None, max_length=16)

    requester_name: str | None = Field(default=None, max_length=255)
    requester_email: str | None = Field(default=None, max_length=255)
    counterparty_name: str | None = Field(default=None, max_length=255)

    due_date: date | None = None
    assigned_to: uuid.UUID | None = None
    linked_contract_id: uuid.UUID | None = None
    linked_template_id: uuid.UUID | None = None

    metadata_json: dict[str, Any] | None = None


class ContractRequestUpdateRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    title: str | None = Field(default=None, min_length=1, max_length=255)
    description: str | None = None

    request_type: str | None = Field(default=None, max_length=64)
    contract_type: str | None = Field(default=None, max_length=64)

    # Constrained to the ContractRequestStatus enum in the route handler.
    status: str | None = Field(default=None, max_length=16)
    priority: str | None = Field(default=None, max_length=16)

    requester_name: str | None = Field(default=None, max_length=255)
    requester_email: str | None = Field(default=None, max_length=255)
    counterparty_name: str | None = Field(default=None, max_length=255)

    due_date: date | None = None
    assigned_to: uuid.UUID | None = None
    linked_contract_id: uuid.UUID | None = None
    linked_template_id: uuid.UUID | None = None

    metadata_json: dict[str, Any] | None = None


class ContractRequestResponse(BaseModel):
    model_config = ConfigDict(extra="forbid", from_attributes=True)

    id: uuid.UUID
    organization_id: uuid.UUID
    title: str
    description: str | None = None
    request_type: str | None = None
    contract_type: str | None = None
    status: str
    priority: str | None = None
    requester_name: str | None = None
    requester_email: str | None = None
    counterparty_name: str | None = None
    due_date: date | None = None
    assigned_to: uuid.UUID | None = None
    linked_contract_id: uuid.UUID | None = None
    linked_template_id: uuid.UUID | None = None
    created_at: datetime
    updated_at: datetime
    created_by: uuid.UUID | None = None
    metadata_json: dict[str, Any] | None = None


# ---------------------------------------------------------------------------
# Request -> Contract conversion
#
# Converting a request runs the same ``AgreementTemplate`` generation
# path the templates surface uses, then links the resulting Contract
# back to the request. The response carries both so the UI can pivot
# straight into the contract workspace without a second round trip.
# ---------------------------------------------------------------------------


class ConvertRequestToContractRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    title: str | None = Field(default=None, max_length=500)
    variable_values: dict[str, Any] = Field(default_factory=dict)


class ConvertRequestToContractResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    request: ContractRequestResponse
    contract: ContractListItemResponse
    artifact: ContractArtifactResponse
    markdown_snapshot: ContractMarkdownSnapshotResponse | None = None
    variables_used: list[str] = Field(default_factory=list)


# ---------------------------------------------------------------------------
# Request -> Contract conversion via uploaded file (PR #65)
#
# The other intake path: a user takes an open ContractRequest and
# uploads an external agreement file (counterparty paper, third-party
# signed exhibit, etc.) instead of generating from a template. The
# uploaded file becomes the Contract's ``original_upload`` artifact and
# the request is linked + completed in the same transaction.
# ---------------------------------------------------------------------------


class ConvertRequestUploadResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    request: ContractRequestResponse
    contract: ContractListItemResponse
    artifact: ContractArtifactResponse
    markdown_snapshot: ContractMarkdownSnapshotResponse | None = None
    # PR #66 — same suggestion + duplicate-warning block the Repository
    # upload returns. Either may be empty; neither blocks the
    # conversion.
    extracted_metadata: ExtractedContractMetadataResponse | None = None
    duplicate_candidates: list[DuplicateContractCandidateResponse] = Field(
        default_factory=list
    )
